CONNECT-UDP: 2021 Design Team Consensus (#77)
ietf-wg-masque/draft-ietf-masque-connect-udp: draft-ietf-masque-connect-udp.md

but not required.

# Context Identifiers {#context-id}

This protocol allows future extensions to exchange HTTP Datagrams which carry
different semantics from UDP payloads. Some of these extensions can augment UDP
payloads with additional data, while others can exchange data that is completely
separate from UDP payloads. In order to accomplish this, all HTTP Datagrams
associated with UDP Proxying request streams start with a context ID, see
{{format}}.

Context IDs are 62-bit integers (0 to 2<sup>62</sup>-1). The context ID value of
zero is reserved for UDP payloads, while non-zero values are dynamically

> allocated: non-zero even-numbered context IDs are client-initiated, and

Author: I prefer the term initiated since that maps to what QUIC uses to refer to streams. If that's OK with you I'm going to declare this editorial and that we can discuss it with the WG instead of in the design team.

> allocated: non-zero even-numbered context IDs are client-initiated, and

Reviewer: That being the case, please be consistent and use "initiate" rather than "allocate" later.  I was just trying to make the terms you use in the definition consistent with the usage later.  (In the direction I prefer, but you do get editor's privilege there.)

> allocated: non-zero even-numbered context IDs are client-initiated, and

Author: Fair enough, you've convinced me. Switched to client-allocated.

> odd-numbered context IDs are server-initiated. Extensions can use the next

Reviewer 2: ```suggestion
odd-numbered context IDs are server-initiated. Extensions can use any
```

I think it could be valuable to register IDs out of order.  Unless we have some compelling reason to require registration in-order, I would prefer to make it clearer that this is allowed.

(I think composition is likely to work by first registering the components, and then constructing the composition using the components by ID.  Since the composition is the one we care about, we don't want to waste short IDs on the components that will not actually be used independently.)

available context ID for their own purposes. Note that, once allocated, any
context ID can be used by both client and server - only allocation carries

> separate namespaces to avoid requiring synchronization. Additionally, note that

Reviewer: Is there any constraints on allocation?  As a client, can I perform random allocations from the space?  (yes?) If I allocate 2^62-1 as a server, does that mean that all previous values are taken? (no?) What if a peer attempts to overload my allocation table? (suck it? we'll not be making it easy to consume multiple allocations anyway ...?)

Author: Added some text to clarify that random allocations are allowed

the context ID namespace is tied to a given HTTP request: it is possible for the
same numeral context ID to be used simultaneously in distinct requests.

# HTTP Datagram Payload Format {#format}

When associated with UDP proxying request streams, the HTTP Datagram Payload
field of HTTP Datagrams (see {{HTTP-DGRAM}}) has the following format:

~~~
UDP Proxying HTTP Datagram Payload {
  Context ID (i),
  Payload (..),

> }

Author: Unfortunately that wouldn't be correct. The Quarter Stream ID is present in the "QUIC DATAGRAM frame" encoding of HTTP Datagrams, but not in the Capsule encoding. Both encodings contain the HTTP Datagram Payload field.

> }

Reviewer: Right.  I did say that maybe you can use words, which would be fine.  The problem here is that I don't quite know how this is expected to fit into the bigger picture without a drawing on a lot more context.  I have that context, but I fear that others will not.

> }

Author: Fair enough, I've added some text right above the diagram.

~~~
{: #dgram-format title="UDP Proxying HTTP Datagram Format"}

Context ID:

: A variable-length integer that contains the value of the Context ID. Endpoints

> MUST silently drop any received HTTP Datagram which carries an unknown Context

Author: That was an oversight, fixed.

This protocol allows future extensions to exchange HTTP Datagrams which carry
different semantics from UDP payloads. Some of these extensions can augment UDP
payloads with additional data, while others can exchange data that is completely
separate from UDP payloads. In order to accomplish this, all HTTP Datagrams
associated with UDP Proxying request streams start with a context ID, see
{{format}}.

Context IDs are 62-bit integers (0 to 2<sup>62</sup>-1). Context IDs are encoded
as variable-length integers, see {{Section 16 of QUIC}}. The context ID value of
0 is reserved for UDP payloads, while non-zero values are dynamically allocated:
non-zero even-numbered context IDs are client-allocated, and odd-numbered
context IDs are server-allocated. Implementations will provide a context ID
allocation service. That service will allow extensions to request a unique
context ID that they can subsequently use for their own purposes. This means
that an HTTP client implementation of the context ID allocation service MUST
only provide even-numbered IDs, while a server implementation MUST only provide
odd-numbered IDs. The context allocation service MUST NOT return the same

> context ID twice, however it MAY return IDs in any order. Once allocated, any

Author: I'm not sure what you're suggesting. Can you elaborate?

> context ID twice, however it MAY return IDs in any order. Once allocated, any

Reviewer 3: https://github.com/ietf-wg-masque/draft-ietf-masque-connect-udp/pull/79

> context ID twice, however it MAY return IDs in any order. Once allocated, any

Author: Merged #79

separate namespaces to avoid requiring synchronization. Additionally, note that
the context ID namespace is tied to a given HTTP request: it is possible for a
context ID with the same numeric value to be simultaneously assigned different
semantics in distinct requests, potentially with different semantics.

Registration is the action by which an endpoint informs its peer of the
semantics and format of a given context ID.

# HTTP Datagram Payload Format {#format}

When associated with UDP proxying request streams, the HTTP Datagram Payload
field of HTTP Datagrams (see {{HTTP-DGRAM}}) has the format in {{dgram-format}}.
Note that when HTTP Datagrams are encoded using QUIC DATAGRAM frames, the
Context ID field below directly follows the Quarter Stream ID field which is at
the start of the QUIC DATAGRAM frame payload.

~~~
UDP Proxying HTTP Datagram Payload {
  Context ID (i),
  Payload (..),
}
~~~
{: #dgram-format title="UDP Proxying HTTP Datagram Format"}

Context ID:

: A variable-length integer that contains the value of the Context ID. If an
HTTP/3 datagram which carries an unknown Context ID is received, the receiver

> SHALL either drop that datagram silently or buffer it temporarily while awaiting

Author: What kind of guidance did you have in mind?

> SHALL either drop that datagram silently or buffer it temporarily while awaiting

Reviewer 3: 1xRTT?

> SHALL either drop that datagram silently or buffer it temporarily while awaiting

Author: Added

of a round trip) while awaiting the registration of the corresponding Context ID.

Payload:

: The payload of the datagram, whose semantics depend on value of the previous
field. Note that this field can be empty.

UDP packets are encoded using HTTP Datagrams with the Context ID set to zero.

> When the Context ID is set to zero, the Context Payload field contains the

Reviewer 4: ```suggestion
When the Context ID is set to zero, the Payload field contains the
```

: This document.

--- back

> # Example Extensions

Reviewer: I feel that these extensions are a little too well formed to serve well as extensions.  They read far too much like specifications, which might be misleading or confusing.

If the goal is to guide future protocol designers, that guidance can be provided as pure text.  Including "grammar" (illustrations really, if I stick to my position there) isn't necessary to convey the necessary meaning.

Author: I don't disagree, but since we're most likely to remove this appendix entirely before publication, I'd rather not spend too much time on it. I've added Example- in more places to make this clearer

Extensions can define new semantics for the payload of HTTP Datagrams. The
extension can then have an endpoint pick an available context ID using the

> allocation service ({{context-id}}) and register that context ID with their

Author: That was also an oversight, fixed.

peer. Registration is the action by which an endpoint informs its peer of the

> semantics and format of a given context ID.

Author: Agreed, moved.

> ## Registering Contexts with Headers

Reviewer 2: I would appreciate some clarification on whether a single extension is expected to offer _both_ registration mechanisms.

If most extensions are expected to support both, I think we should consider adding some standards development guidance to minimize duplicated effort (e.g. registration capsule contents should be exactly the header value).

Reviewer 5: I think some guidance would be useful but it would make most sense to put it alongside the Context ID definition in https://ietf-wg-masque.github.io/draft-ietf-masque-connect-udp/2021_design_team_consensus/draft-ietf-masque-connect-udp.html#section-4

I don't think that needs to block this PR but can instead be done as a follow up WG item.

Author: During the last design team meeting, we reached consensus that extension composability should be addressed by the working group as a whole, instead of as part of the design team efforts. When we do look into this as a working group, I'm fairly confident that the appendix about extensions will get rewritten. That'll be the best time to discuss what guidance we want to provide in terms of whether a given extension should offer both registration mechanisms. All this to say I agree with Lucas, let's handle this as a follow up WG item - @bemasc feel free to file an issue to make sure this is tracked.

Extensions can define a new HTTP header to register a context ID with the peer
endpoint.

> As an example, let's use an extension that conveys the time at which a UDP

Reviewer: House style doesn't use "us" or "we".

```suggestion
As an example, take an extension that conveys the time at which a UDP
```

packet was received. The extension would first define the format of its HTTP
Datagram Context Payload field:

~~~
Context Payload for UDP with Timestamp {
  Timestamp (64),
  UDP Payload (..),
}

> ~~~

Reviewer: This definitely needs to include the context ID.  As an example, the more context (no pun intended) the better.

Author: Agreed, done.

{: #ex-dgram title="Example: Format of UDP Payload with Timestamp"}

> The extension would also define a new HTTP header (UDP-Timestamps) that includes

Reviewer: It might pay to prefix this with "Example-" in order to (maybe) avoid a potential misunderstanding about whether this is a real thing.

Author: Agreed, done.

a context ID value. Endpoints that understand this new HTTP header would be able
to consequently handle and parse datagrams on the context ID, while all other

> endpoints would ignore the datagrams.

Author: I've clarified the text about this specific client-initiated extension example.

However, we don't need to do anything else for unidirectional extensions. Such an extension could easily say that its registration is unidirectional, and that sending in the other direction is not allowed. There's no reason to split the namespace - QUIC did that because a bidirectional stream has a more complex state machine, and bidirectional flow control - neither of which apply to datagrams.

~~~
HEADERS
:method = CONNECT
:protocol = connect-udp
:scheme = https
:path = /192.0.2.42/443/
:authority = proxy.example.org

> udp-timestamps = 42

Author: We decided to punt registration and extension composability back to the WG

timestamp.

## Registering Contexts with Capsules

Extensions can define a new Capsule type (see {{HTTP-DGRAM}}) to register a
context ID with the peer endpoint.

As an example, let's use an extension that compresses QUIC Connection IDs when
the client is running QUIC over a UDP proxying tunnel. The extension would first
define the transform applied to UDP payloads when compressing and decompressing,
such as removing the bytes of the connection ID.

The extension would also define a new capsule type
(REGISTER_COMPRESSED_QUIC_CID) that includes a context ID value and the
connection ID to compress. Endpoints that understand this new capsule type would
be able to consequently handle and parse datagrams on the context ID, while all
other endpoints would ignore the datagrams.

~~~
REGISTER_COMPRESSED_QUIC_CID Capsule {
  Type (i) = REGISTER_COMPRESSED_QUIC_CID,
  Length (i),
  Context ID (i),
  QUIC Connection ID (..),
}
~~~
{: #ex-capsule title="Example: Registration via capsule"}

## Composing Extensions

A future extension could define how to compose extensions. For example, it could
define a new HTTP header and/or capsule that maps a context ID to a list of

> extensions. That allows the sender to pick low context IDs for the combinations

Reviewer: I find this "low context IDs" thing to be a little bit of a non-sequitur.  You haven't established why composition might be useful or important, but now you are jumping straight to optimizing for size.  This section probably needs a bit more context.

Author: I've removed this subsection. We'll decide as a WG whether we want to tackle extension composition in this draft or not

: This document.

--- back

# Example Extensions

Extensions can define new semantics for the payload of HTTP Datagrams. The
extension can then have an endpoint pick an available context ID using the
allocation service ({{context-id}}) and register that context ID with their
peer.

## Registering Contexts with Headers

Extensions can define a new HTTP header to register a context ID with the peer
endpoint.

As an example, take an extension that conveys the time at which a UDP
packet was received. The extension would first define the format of its HTTP
Datagram Payload field:

~~~
UDP with Timestamp HTTP Datagrams {
  Context ID (i),
  Timestamp (64),
  UDP Payload (..),
}
~~~
{: #ex-dgram title="Example: Format of UDP Payload with Timestamp"}

The extension would also define a new HTTP header (Example-UDP-Timestamps) that
includes a context ID value. Servers that understand this new HTTP header would
be able to consequently handle and parse datagrams on the context ID, while all
other servers would silently drop the datagrams. Clients also indicate their

> support for receiving on this Context ID by registering it.

Author: 1. I personally don't think there's value in an acknowledgement, but I think that's something we'll need to hash out with the WG when we discuss extension composability and registration
2. In general, yes. Individual extensions can prohibit it though
3. Not in this particular example, but an extension can absolutely do that

At a high level, this appendix is most likely not going to survive publication, let's see what it looks like once we've figured out registration and composability with the WG

> support for receiving on this Context ID by registering it.

Reviewer 3: I'm not really comfortable with that, because it means we'll just rediscover these points.

If we don't want to fix up this appendix, let's move it into a wiki or whatever.

> support for receiving on this Context ID by registering it.

Author: I've added an explicit note stating that this appendix will be removed before WGLC

> support for receiving on this Context ID by registering it.

Reviewer 3: This would not be my preference. If the point of the appendix is to help make this clearer, then having it be confusing does not achieve that.

> support for receiving on this Context ID by registering it.

Author: Fair enough, I've added a paragraph that answers your three questions above

~~~
HEADERS
:method = CONNECT
:protocol = connect-udp
:scheme = https
:path = /192.0.2.42/443/
:authority = proxy.example.org
example-udp-timestamps = 42
~~~
{: #ex-hdr title="Example: Registration via header"}

In this example request, HTTP Datagrams with context ID zero would only contain
the UDP payload, whereas HTTP Datagrams with context ID 42 would also contain a
timestamp.

## Registering Contexts with Capsules

Extensions can define a new Capsule type (see {{HTTP-DGRAM}}) to register a
context ID with the peer endpoint.

As an example, take an extension that compresses QUIC Connection IDs when the
client is running QUIC over a UDP proxying tunnel. The extension would first
define the transform applied to UDP payloads when compressing and decompressing,
such as removing the bytes of the connection ID.

The extension would also define a new capsule type
(EXAMPLE_REGISTER_COMPRESSED_QUIC_CID) that includes a context ID value and the
connection ID to compress. Endpoints that understand this new capsule type would
be able to consequently handle and parse datagrams on the context ID, while all

> other endpoints would ignore the datagrams.

Reviewer 3: Again, I see that this is an example, but I don't see how this works as a practical matter. What acknowledges this? How do I discover that the peer doesn't support this extension? Or does, for that matter? I can't just register a compressed CID and then wait to see if my  data goes through.

Author: Added a sentence about needing a header
